Connection Dialog: Dropdown for database name

abist opened the pull request:
Instead of iterating on my previous PR here - https://github.com/Microsoft/sqlopsstudio/pull/489, I decided to make a new pull request altogether.

Changes from previous iteration/existing Connection Dialog -

1. Now uses an editable dropdown. 

![dropdown2-1](https://user-images.githubusercontent.com/6411451/35543167-861b1452-0518-11e8-8553-cd791079f0b5.PNG)

2. The UI is **not** frozen any more when a connection is made. The dropdown just shows "Loading..." until the dropdown options are populated by the list database task.

![dropdown2-2](https://user-images.githubusercontent.com/6411451/35543224-c99c89c2-0518-11e8-9024-8cc3e2ada893.PNG)

3. When a connection is selected from "Recent History", the database dropdown is still functional and doesn't have the database name set.

![dropdown2-3](https://user-images.githubusercontent.com/6411451/35543312-4f203012-0519-11e8-9241-584e52e29a5a.PNG)

4. Supports SQL Login in addition to Windows Integrated Auth.

![dropdown2-4](https://user-images.githubusercontent.com/6411451/35543388-b08202cc-0519-11e8-9942-cef25c01790f.PNG)

MattIrv commented on src/sql/parts/connection/connectionDialog/connectionWidget.ts:
```diff
@@ -238,6 +241,31 @@ export class ConnectionWidget {
+			}
+		}));
+
+		this._toDispose.push(this._databaseNameInputBox.onValueChange(s => {
```
minor, but I'd suggest using a more descriptive variable name than `s`

MattIrv commented on src/sql/parts/connection/connectionDialog/connectionController.ts:
```diff
@@ -43,11 +43,39 @@ export class ConnectionController implements IConnectionComponentController {
+						if (result && result.databaseNames) {
+							resolve(result.databaseNames);
+						} else {
+							reject();
```
Not sure when you would get into this situation, but should there be an error message here? Based on the error message being included below I assume it gets shown to the user when this rejects?

abist replied:
This error gets swallowed. The reason this error message isn't shown to the user because it takes a lot of time to show up the error message, and ruins the UX.

MattIrv commented on src/sql/parts/connection/connectionDialog/connectionController.ts:
```diff
@@ -43,11 +43,39 @@ export class ConnectionController implements IConnectionComponentController {
+						}
+					})
+				} else {
+					reject(connResult.errorMessage);
```
What's the drop down's behavior when this rejects? Is the user still shown the "default" option and able to type in another option?

abist replied:
Yes

kburtram commented:
@abist it looks like the drop down list can go off the bottom of the screen without a scrollbar.  We should limit the size of the drop down to the bottom of the shell window and provide a scrollbar.  Please see below screenshot about what I'm talking about.  Notice the last item is clipped without a scrollbar.

![screen](https://user-images.githubusercontent.com/599935/35703103-de81352a-074f-11e8-8dcd-9cd002cdfadb.png)

kburtram commented:
@abist I noticed that the dropdown filtering is case-sensitive.  For example 'a' filters out 'AdventureWorks'.  I think we probably be case-insensitive while filtering, but when the user selects an item use the casing provided by server.  For example, user types 'adv' and sees 'AdventureWorks' selects the item and final text is 'AdventureWorks'

kburtram commented:
@abist it looks like each time I expand the drop down it does a "Loading" refresh.  I wonder if we should cache the results client side if server & credentials don't change.  This optimizes for slower servers with lots of DBs where populating the list can take a long time (~30 seconds or more).

kburtram commented:
@abist it looks like the drop down list items are vertically top aligned.  I think the items should be vertically centered, or not have so much bottom padding.  Please checkout the other drop-down boxes in the dialog for examples.

MattIrv commented on src/sql/parts/connection/connectionDialog/connectionWidget.ts:
```diff
@@ -238,6 +241,31 @@ export class ConnectionWidget {
 		}));
+
+		this._toDispose.push(this._databaseNameInputBox.onFocus(() => {
+			if (this.serverName) {
```
It might be best to do this only when the database list isn't already loaded. Right now it appears to reload every time I click on the database selection - not too bad connecting to my local server where the list comes back instantly, but could be annoying on a slow connection

abist replied:
Added a cache to eliminate this problem

abist commented:
@kburtram Addressed your comments. Changed the height of the dropdown, made the filter case-insensitive, aligned the text for the dropbox and added a cache for connections made for the database list.

kburtram commented:
@abist do you know why the dropdown arrow is styled differently than the other dropdowns in the dialog?  I think @AbbiePetcht looked at a similar issue recently.

![screen](https://user-images.githubusercontent.com/599935/35884978-3f998cc8-0b41-11e8-803f-7fd6d40ea145.png)

abist commented:
@kburtram This is the issue I mentioned in the design email. The editable dropdown `Dropdown` class that I just added is different from the other dropdown `SelectBox` class, that the connection dialog is using. @AbbiePetcht mentioned that we have this inconsistency elsewhere as well.

kburtram commented:
@abist the alignment and spacing of the list box still doesn't look correct to me.  Please look at the styling of any other dropdown on the dialog for example (such as Authentication Type, this control should be vertically aligned and top\bottom padding the same).  Also, the list is still going off the bottom of the screen without a scrollbar as mentioned previously.

![screen2](https://user-images.githubusercontent.com/599935/35885132-be90983c-0b41-11e8-9cdd-721ecac755cd.png)


For reference here's how the vertical alignment and padding should look.

![screen3](https://user-images.githubusercontent.com/599935/35885221-0a5a8520-0b42-11e8-9465-26a9df52c030.png)

kburtram commented:
@abist fyi, I received the below error testing this feature.  I don't have a repro beyond playing with the feature for a bit.

![screen4](https://user-images.githubusercontent.com/599935/35885301-43a8217a-0b42-11e8-9ff9-7aec1c5e2895.png)

kburtram commented:
@abist regarding the UI inconsistency mentioned above.  Why can't we fix this style to be consistent?  Could you please open issues for the inconsistency in other places in the UI?

AbbiePetcht commented:
The selectbox is a native dropdown while the editable dropdown is the custom dropdown, so it will look very different on the mac. We already have the inconsistent look in restore dialog. If we want the consistent look we need to use the custom dropdown for all dropdowns in flyouts. For windows, I think you can try to make arrow to look similar to the selectbox.

---
In reply to: [363569742](https://github.com/Microsoft/sqlopsstudio/pull/583#issuecomment-363569742) [](ancestors = 363569742)

kburtram commented:
@AbbiePetcht we should optimize for Windows if we need to choose a particular platform (since that's where most the users are), assuming it doesn't look bad on mac and Linux.  Can we make it look consistent on Windows, or will this be a lot of work?

kburtram commented:
@abist I also just noticed the database list isn't sorted.  Can we please sort the list alphabetically to make it easier to locate items when there are lots of DBs?

kburtram commented:
@abist filtering the list is still case-sensitive as previously mentioned.  Please go back and review all comments on the PR since most of the previous comments seem unaddressed in my testing as far as I can tell.  Thanks!

AbbiePetcht commented:
@abist Please make the css in editableDropdown to look similar to selectBox

abist commented:
@kburtram I think you might have the old (first iteration only). From your screenshot, the dropdown size hasn't changed and the alignment is unchanged, as you mentioned. The case-sensitive filtering isn't working for you either, so my best guess is to try pulling again.

abist commented:
@AbbiePetcht @kburtram  I don't think we can change the icon for `Dropdown` to be the same as `SelectBox`, because the dropdown icon that's being rendered for `SelectBox` is from the native HTML `select` class. What we can do is change the `SelectBox` arrow to be like `Dropdown`'s, but it may be a bit of work.

kburtram commented:
@abist I'll try to merge your branch again to see if I get the latest fixes, though I merged `connDialog/dropdown-2` into the latest `master` when testing today. 

Regarding drop down style, the request isn't to use the exact same icon as the select box or to globally replace select box.  The request is much more scoped, please update the drop down icon to look as close as reasonably possible to the select box icon on Windows platform.  Basically make it a bit more narrow and position it more to the right.  Please let me know if I'm missing something and this request is more difficult than it seems.  Thanks!

kburtram commented:
@abist you're right, my test branch was missing a couple of your latest commits.  It looks like case-insensitive filtering is working.

The drop-down arrow and vertical alignment issues are still there.  I assume due to differences between SELECT and custom drop down controls.  Though as mentioned above we should try to make custom drop down look as much like Windows SELECT as reasonable without significantly compromising macOS and Linux UX.

Also, I'm hitting several "keyboard only" issues.  Like tab doesn't select the best match (have to down arrow to desired it) and ESC does dismiss the drop down, but closes whole dialog.  Please try to use the Connection Dialog without using the mouse at all and make sure this control is functional and intuitive.

I'd recommend that we take this PR as-is so it gets included in tonight's test build.  Then take the UX and keyboard navigation issues separately.  I don't think those issues are "ship blockers" for this scenario for Feb.
